boskos/cmd/cli/cli.go

```diff
@@ -0,0 +1,129 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+	"k8s.io/test-infra/boskos/client"
+	"k8s.io/test-infra/boskos/common"
+)
+
+type options struct {
+	// common, used to create the client
+	serverURL string
+	ownerName string
+
+	c *client.Client
+
+	acquire acquireOptions
+}
+
+type acquireOptions struct {
+	requestedType  string
+	requestedState string
+	targetState    string
+	wait           bool
+}
+
+func main() {
+	options := options{}
+
+	root := &cobra.Command{
+		Use:   "boskosctl",
+		Short: "Boskos command-line client for resource leasing",
+		Long: `Boskos provides a flexible resource leasing server.
+
+The boskosctl is a command-line client for this server,
+allowing for a user to acquire and release leases from
+scripts with a simple interface.`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			// the root command does nothing, so just print help
+			return cmd.Help()
+		},
+		Args: cobra.NoArgs,
+	}
+	root.PersistentFlags().StringVar(&options.serverURL, "server-url", "", "URL of the Boskos server")
+	root.PersistentFlags().StringVar(&options.ownerName, "owner-name", "", "Name identifying the user of this client")
+	for _, flag := range []string{"server-url", "owner-name"} {
+		if err := root.MarkPersistentFlagRequired(flag); err != nil {
+			fmt.Println(err)
+			os.Exit(1)
+		}
+	}
+
+	acquire := &cobra.Command{
+		Use:   "acquire",
+		Short: "Acquire resource leases",
+		Long: `Acquire a resource lease, either best-effort or blocking.
+
+Resources can be leased by identifying which type of resource is needed
+and what state the resource should be in when leased. Resources will also
+transition to a new state upon being leased.
+
+On a successful lease acquisition, the leased resource will be printed in
+JSON format for downstream consumption.
+
+Examples:
+
+  # Acquire one clean "my-thing" and mark it dirty when leasing
+  $ boskosctl acquire --type my-thing --state clean --target-state dirty
+
+  # Acquire one new "my-thing" and mark it old when leasing, block until successfully leased
+  $ boskosctl acquire --type my-thing --state new --target-state old --wait`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			options.c = client.NewClient(options.ownerName, options.serverURL)
+			acquireFunc := options.c.Acquire
+			if options.acquire.wait {
+				acquireFunc = func(rtype, state, dest string) (resource *common.Resource, e error) {
+					return options.c.AcquireWait(context.Background(), rtype, state, dest)
+				}
+			}
+			resource, err := acquireFunc(options.acquire.requestedType, options.acquire.requestedState, options.acquire.targetState)
```

Comment:
do you also want to handle retry here? (it's currently hard-coded in the client: https://github.com/kubernetes/test-infra/blob/master/boskos/client/client.go#L83)

Reply:
If we refactor the client constructor to expose it, this can be a follow-up. I want the same UX right now for the CLI as we have for the client library users